global/scripts/debug.shlib

```diff
@@ -0,0 +1,275 @@
+then
+    source "$HCPPIPEDIR/global/scripts/log.shlib" "$@"
+fi
+
+HCP_DEBUG_USETRAP=ON
+HCP_DEBUG_EXIT_ON_ERROR=ON
+
+if [ -z ${HCP_DEBUG_COLOR} ] ; then
+    HCP_DEBUG_COLOR=FALSE
+fi
+
+if [ -z ${HCP_VERBOSE} ] ; then
+    HCP_VERBOSE=FALSE
+fi
+
+# Description:
+#
+#   Function to turn on error trapping
+
+debug_enable_trap()
+{
+    HCP_DEBUG_USETRAP=ON
+}
+
+# Description:
+#
+#   Function to turn off error trapping
+
+debug_disable_trap()
+{
+    HCP_DEBUG_USETRAP=OFF
+}
+
+# Description:
+#
+#   Function to turn on exit upon error trap
+
+debug_enable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=ON
+}
+
+# Description:
+#
+#   Function to turn off exit upon error trap
+
+debug_disable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=OFF
+}
+
+# Description:
+#
+#   Function to turn on output coloring
+
+debug_enable_color()
+{
+    HCP_DEBUG_COLOR=TRUE
+}
+
+# Description:
+#
+#   Function to turn off output coloring
+
+debug_disable_color()
+{
+    HCP_DEBUG_COLOR=FALSE
+}
+
+
+# Description:
+#
+#   Function for printing red debug output.
+#
+# Input:
+#
+#   - text to be printed
+
+debug_red_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[31m$*\033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing green debug output.
+#
+# Input:
+#
+#   - text to be printed
+
+debug_green_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[32m$*\033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+
+# Description:
+#
+#   Function for printing blue debug output.
+#
+# Input:
+#
+#   - text to be printed
+
+debug_blue_echo()
+{   
+    if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+        echo -e "\033[34m$*\033[0m"
+    else
+        echo "$*"
+    fi
+}
+
+function debug_first()
+{
+    echo "$1"
+}
+
+debug_on_error()
+{    
+    if [ ! $HCP_DEBUG_USETRAP == 'OFF' ]; then        
+
+        local me=`basename "$0"`
+        local startreport=""
+        local endreport=""
+        local command=$(eval debug_first "$2")
+
+        # --- output a regular HCP error log
+
+        log_Err "'$command' command failed with return code: $3"
+
+        # --- are we printing in color
+
+        if [ "$HCP_DEBUG_COLOR" == "TRUE" ] ; then
+           startreport="\033[33m"
+           endreport="\033[0m"
+        fi
+        
+        # --- output a detailed report if usetrap is not turned off    
+        echo                                                               
+        echo -e "${startreport}===> ERROR: Command returned with nonzero exit code"
+        echo "---------------------------------------------------"
+        echo "         script: $me"                                            
+        echo "stopped at line: $1"                                    
+        echo "           call: $2"
+        echo "  expanded call: $(eval echo "$2")"                            
+        echo "    exit status: $3"                                        
```

Comment:
Minor quibble, but we alternatively refer to this as "return code" (in the `log_Err` message), "nonzero exit code" in the detailed report, and "exit status" here.

Reply:
BTW: With my suggestions/comments this morning, I believe that I'm truly done with this PR, unless feedback on those suggestions/comments requires a further response.

Reply:
I'll set it to `exit code` as it is obvious that it is nonzero.